pkg/util/selector_test.go

```diff
@@ -297,313 +297,372 @@ func TestClusterMatches(t *testing.T) {
+					MatchLabels: map[string]string{"foo": "unmatched"},
+				},
+				ClusterNames: []string{},
+				FieldSelector: &policyv1alpha1.FieldSelector{
+					MatchExpressions: []corev1.NodeSelectorRequirement{
+						{Key: ZoneField, Operator: corev1.NodeSelectorOpExists},
 					},
 				},
 			},
 			want: false,
 		},
 		{
-			name: "[case 1] clusterNames not matched",
-			args: args{
-				affinity: policyv1alpha1.ClusterAffinity{
-					LabelSelector: &metav1.LabelSelector{
-						MatchLabels: map[string]string{"foo": "bar"},
-					},
-					ClusterNames: []string{"unmatched"},
-					FieldSelector: &policyv1alpha1.FieldSelector{
-						MatchExpressions: []corev1.NodeSelectorRequirement{
-							{Key: ZoneField, Operator: corev1.NodeSelectorOpExists},
-						},
+			name: "test cluster names not matched",
+			affinity: policyv1alpha1.ClusterAffinity{
+				LabelSelector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{"foo": "bar"},
+				},
+				ClusterNames: []string{"unmatched"},
+				FieldSelector: &policyv1alpha1.FieldSelector{
+					MatchExpressions: []corev1.NodeSelectorRequirement{
+						{Key: ZoneField, Operator: corev1.NodeSelectorOpExists},
 					},
 				},
 			},
 			want: false,
 		},
 		{
-			name: "[case 1] fields not matched",
-			args: args{
-				affinity: policyv1alpha1.ClusterAffinity{
-					LabelSelector: &metav1.LabelSelector{
-						MatchLabels: map[string]string{"foo": "bar"},
-					},
-					ClusterNames: []string{cluster.Name},
-					FieldSelector: &policyv1alpha1.FieldSelector{
-						MatchExpressions: []corev1.NodeSelectorRequirement{
-							{Key: ProviderField, Operator: corev1.NodeSelectorOpExists},
-						},
+			name: "test field selector provider matched",
+			affinity: policyv1alpha1.ClusterAffinity{
+				LabelSelector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{"foo": "bar"},
+				},
+				ClusterNames: []string{cluster.Name},
+				FieldSelector: &policyv1alpha1.FieldSelector{
+					MatchExpressions: []corev1.NodeSelectorRequirement{
+						{Key: ProviderField, Operator: corev1.NodeSelectorOpExists},
 					},
 				},
 			},
-			want: false,
+			want: true,
 		},
 		{
-			name: "[case 1] matched",
-			args: args{
-				affinity: policyv1alpha1.ClusterAffinity{
-					LabelSelector: &metav1.LabelSelector{
-						MatchLabels: map[string]string{"foo": "bar"},
-					},
-					ClusterNames: []string{cluster.Name},
-					FieldSelector: &policyv1alpha1.FieldSelector{
-						MatchExpressions: []corev1.NodeSelectorRequirement{
-							{Key: ZoneField, Operator: corev1.NodeSelectorOpExists},
-						},
+			name: "test field selector zone matched",
+			affinity: policyv1alpha1.ClusterAffinity{
+				LabelSelector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{"foo": "bar"},
+				},
+				ClusterNames: []string{cluster.Name},
+				FieldSelector: &policyv1alpha1.FieldSelector{
+					MatchExpressions: []corev1.NodeSelectorRequirement{
+						{Key: ZoneField, Operator: corev1.NodeSelectorOpExists},
```

Comment:
`NodeSelectorOpExists` is invalid. We should use `NodeSelectorOpIn`.

Reply:
This case has existed before. I understand that the test here is to check if zone exists

Reply:
Yeah, it is exactly why we need to refine the test.

For this case, we want to test that a field selector could be used to select some clusters, so we have to use a `valid` match expression.
Using the invalid operator here is more like demonstrating the bug of `ClusterMatches`, but this isn't a bug because of #3193.